das/state.go

```diff
@@ -1,53 +1,234 @@
 package das
 
 import (
-	"sync"
-	"time"
+	"context"
 )
 
-// state collects information about the DASer process. Currently, there are
-// only two sampling routines: the main sampling routine which performs sampling
-// over current network headers, and the `catchUp` routine which performs sampling
-// over past headers from the last sampled checkpoint.
-type state struct {
-	sampleLk sync.RWMutex
-	sample   RoutineState // tracks information related to the main sampling routine
+// represents current state of sampling
+type coordinatorState struct {
+	rangeSize uint64
 
-	catchUpLk sync.RWMutex
-	catchUp   JobInfo // tracks information related to the `catchUp` routine
+	priority   []job                      // list of headers heights that will be sampled with higher priority
+	inProgress map[int]func() workerState // keeps track of running workers
+	failed     map[uint64]int             // stores heights of failed headers with amount of attempt as value
+
+	nextJobID int
+	next      uint64 // all headers before next were sent to workers
```

Comment:
should next not be protected somehow or read from / written to atomically?

Reply:
same thing with `priority`

Reply:
Unless it's guaranteed reads on `coordinatorState` can never happen at the same time as `nextJob` determination?

Reply:
> corrdinator is not blocked from updating maxKnown or exporting stats to any external request.
This piece of code is responsible for preventing race condition on collecting stats from corrdinator. It blocks coordinator with waitGroup untill stats has been read and unblocks after by releasing waitGroup. This operation is also running on demand, while coordinator is not busy starting new workers or updating state.

`nextJob` will never be called simultaneously as stats are collected?

Reply:
Yes. No need for protection, because it is protected by ownership scope.